jax/experimental/mosaic/gpu/core.py

```diff
@@ -130,8 +139,13 @@ def _mosaic_gpu_lowering_rule(
       operands=args,
       operand_layouts=[list(reversed(range(a.ndim))) for a in ctx.avals_in],
       result_layouts=[list(reversed(range(a.ndim))) for a in ctx.avals_out],
-      backend_config=kernel_id + module_asm,
+      backend_config=dict(
+          kernel_hash=ir.StringAttr.get(kernel_id),
+          module=ir.StringAttr.get(module_asm),
+          use_custom_barrier=ir.BoolAttr.get(use_custom_barrier),
+      ),
```

Comment:
I have a bit of a concern with changing the format of backend config here, because it will affect serialization stability. This is the first time I'd be using those FFI APIs and I will have to check if they will allow us to evolve this API in a backwards-compatible way.

This change will sadly break backwards compatibility already for any StableHLO artifacts that do export Mosaic kernels. I will have to check in with some internal users before we can safely merge this

Reply:
I kept the new FFI API since it will be used to implement `get_barrier_semaphore()` on GPUs in a following PR.

Reply:
Ok this is a bit annoying, but we won't be able to make the transition to the new FFI API atomically. I think you'll need to (1) check if `ctx.is_forward_compat` and if yes use the old lowering, (2) give this custom call a new name (perhaps `mosaic_gpu_v2`), (3) keep the old C++ path around, and only use the improved one for `mosaic_gpu_v2`.

Reply:
Sounds good.
CC @trevor-m since we need to color buffers in XLA for "mosaic_gpu_v2" custom calls on top of "mosaic_gpu".